spec/queries/patient_summary_query_spec.rb

```diff
@@ -0,0 +1,50 @@
+require "rails_helper"
+
+RSpec.describe PatientSummaryQuery do
+  let(:facility) do
+    create(:facility)
+  end
+
+  let(:really_overdue_appointments) do
+    create_list(:appointment, 2, facility: facility, scheduled_date: 380.days.ago, status: "scheduled")
+  end
+
+  it "returns nothing for no data" do
```

Comment:
Could we use contexts to capture these different scenarios? That'll let us set up the necessary test data in each context in isolation:

```ruby
subject(:patients) { described_class.call(relation: relation, filters: filters) }

let(:filters) { [] }
let(:relation) { PatientSummary.all }

context "when filtering for less than one year overdue"
  let(:filters) { ["only_less_than_year_overdue"] }
  before do
    # make some patients
  end
  it "does some stuff" do
    # assertions
  end  
end

context "when filtering for no phone numbers"
  let(:filters) { ["no_phone_number"] }
  before do
    # make some patients
  end
  it "does some stuff" do
    # assertions
  end  
end
```

Reply:
I find this type of over-extracted RSpec style to be detrimental to reading and learning from the specs, especially when future me or other future devs are trying to read a spec months or years after shipping the original feature.

The guidelines laid out in [An opinionated guide to readable RSpec](https://medium.com/selleo/an-opinionated-guide-to-readable-rspec-part-1-of-2-fe1dce79a478) largely align with what I've settled upon when writing readable specs, in particular this bit:

> Keep everything in “it”

> This is where the readability oriented approach differs the most from the optimization oriented one. The key is to keep most, if not all code required to execute test within individual it block. This means that we should drop all usages of let, subject and described_class and significantly limit the usage of both filters (before, around and after) and shared contexts / examples. This means more code will be duplicated with tests setup growing. However, the key advantage is, that while reading a test, developer will be fully aware of the context and will immediately see how test setup affects assertions. With heavily optimized specs code it might require a lot of navigation within spec file and external files to fully understand the environment in which assertions are checked.

A tldr version of this: the more I have to jump between different abstractions to grok a spec, the less useful that spec is long term for the team.  I like to keep the relevant factories, calling code, and expectations inside the spec itself.  If that gets very difficult to do, that is often a sign that the code under test is doing too much or not the right abstractions, and needs to be rethought and refactored.

Granted, there are certainly cases where a common `before` or `let` is used across 95% of a spec, and is worth extracting out. For the most _relevant_ pieces of the test, though, I like to avoid extraction or DRYing things up.

I'll now get off my soapbox here 😏 . Would be happy to discuss on a zoom call too with ya to see what you think.

Reply:
Agree totally with this.

Reply:
Thanks for the soapbox sesh 😅 My rspec views are certainly a bit traditional (I bet they're 3-4 years old). I've never tried this approach while writing specs. I think the other server devs favor this approach as well, so I'm definitely down to try it.

In that case, I have only one request: Please point this out in my PRs too so I can get into this habit 😄

Reply:
> In that case, I have only one request: Please point this out in my PRs too so I can get into this habit 😄

👍 will do 😀